File: tests/java.pl
#!/usr/bin/perl

use File::pushd;

my $TMPFILE="out-java.$$.tmp";

print "Checking the Java configuration\n";

print STDERR "Java ciphersuites per policy\n";

system("javac tests/java/CipherList.java 1>&2");
if ($? != 0) {
	exit 77;
}

my $dir = 'tests/outputs';

opendir(DIR, $dir) or die $!;

my @javapolicies
    = grep {
        /java\.txt/       # ends in java.txt
        && -f "$dir/$_"   # and is a file
    } readdir(DIR);


foreach my $policyfile (@javapolicies) {
	my $policy = $policyfile;
	$policy =~ s/-[^-]+$//;

	print "Checking policy $policy\n";

	{
		my $pushdir = pushd('tests/java');

		#catch errors in this script now, since the -D option will ignore
		#missing files.
		if (!-e "../../$dir/$policyfile") {
			print "Policy file ../../$dir/$policyfile missing\n";
			exit 1;
		}
		system("java -Djava.security.disableSystemPropertiesFile=true -Djava.security.properties=\"../../$dir/$policyfile\" CipherList >../../$TMPFILE");
	}

	my $lines=`cat $TMPFILE|wc -l`;
	if ("$policy" eq "EMPTY" or "$policy" eq "GOST-ONLY") {
		if ($lines >= 2) { # we allow the SCSV
			print "Empty policy has ciphersuites!\n";
			exit 1;
		}
	} else {
		system("grep \"TLS_EMPTY_RENEGOTIATION_INFO_SCSV\" $TMPFILE >/dev/null 2>&1");
		
		if ($? != 0) {
			print "Could not find TLS_EMPTY_RENEGOTIATION_INFO_SCSV in $policy\n";
			system("cat $TMPFILE");
			exit 1;
		}

		if ($lines <= 1) {
			print "Policy $policy has no ciphersuites!\n";
			system("cat $TMPFILE");
			exit 1;
		}
	}
	system("cat $TMPFILE 1>&2");
	unlink($TMPFILE);
}

exit 0;

File: tests/java.py
#!/usr/bin/python3

import subprocess
import sys
from pathlib import Path

print('Checking the Java configuration')

subprocess.run(['javac', 'tests/java/CipherList.java'],  # noqa: S607
               check=True)

for policy_path in Path('tests', 'outputs').glob('*-java.txt'):
    policy = policy_path.name.removesuffix('-java.txt')

    print(f'Checking policy {policy}')

    # catch errors here, in this script,
    # since the -D option will ignore missing files.

    if not policy_path.exists():
        print(f'Policy file {policy_path} missing')
        sys.exit(1)

    try:
        p = subprocess.run(['java',  # noqa: S607
                            '-Djava.security.disableSystemPropertiesFile=true',
                            f'-Djava.security.properties={policy_path}',
                            '-cp', 'tests/java', 'CipherList', '-l'],
                           check=True,
                           encoding='utf-8',
                           stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError as e:
        print(f'CipherList error {e.returncode}:', file=sys.stderr)
        print(e.stdout, file=sys.stderr)
        sys.exit(1)

    out = p.stdout.rstrip()
    lines = out.split('\n')
    line_count = out.count('\n')

    if policy in {'EMPTY', 'GOST-ONLY'}:
        if line_count >= 2:  # we allow SCSV  # noqa: PLR2004
            print('Empty policy has ciphersuites!', file=sys.stderr)
            print(p.stdout, file=sys.stderr)
            sys.exit(1)
    else:
        if 'TLS_EMPTY_RENEGOTIATION_INFO_SCSV' not in lines:
            print('Could not find TLS_EMPTY_RENEGOTIATION_INFO_SCSV '
                  f'in {policy}', file=sys.stderr)
            print(p.stdout, file=sys.stderr)
            sys.exit(1)

        if line_count <= 1:  # SCSV
            print(f'Policy {policy} has no ciphersuites!', file=sys.stderr)
            print(p.stdout, file=sys.stderr)
            sys.exit(1)
